File: next-main/src/main/java/com/next/reservations/core/service/ReservationTimeManagingService.java
/*
package com.next.reservations.core.service;

import com.next.reservations.core.domain.Reservation;
import com.next.reservations.core.domain.ReservationStatus;
import com.next.reservations.core.domain.ReservationTime;
import com.next.reservations.core.domain.ReservationTimeConfiguration;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class ReservationTimeManagingService {

    private ReservationTimeService reservationTimeService;
    private ReservationTimeConfigurationService reservationTimeConfigurationService;
    private ReservationService reservationService;

    public ReservationTimeManagingService(ReservationTimeService reservationTimeService, ReservationTimeConfigurationService reservationTimeConfigurationService, ReservationService reservationService) {
        this.reservationTimeService = reservationTimeService;
        this.reservationTimeConfigurationService = reservationTimeConfigurationService;
        this.reservationService = reservationService;
    }

    public List<ReservationTime> findUnreservedTimesForDate(LocalDate date, ReservationTimeConfiguration timeConfiguration) {
        final List<Long> reservedTimeIds = reservationService.findAllByStatusInAndReservationDate(List.of(ReservationStatus.ACCEPTED, ReservationStatus.PENDING), date)
                .stream().map(it -> it.getReservationTime().getId()).collect(Collectors.toList());

        final List<ReservationTime> availableTimes = reservationTimeService.findAllByReservationTimeConfigId(timeConfiguration.getId());

        return availableTimes.stream().filter(it -> !reservedTimeIds.contains(it.getId())).collect(Collectors.toList());
    }

    public ReservationTimeConfiguration findConfigurationForDate(LocalDate date) {
        final Optional<ReservationTimeConfiguration> futureDefault = reservationTimeConfigurationService.findFutureDefault();

        if (futureDefault.isPresent() && !date.isBefore(futureDefault.get().getDefaultStartDate())) {
            return futureDefault.get();
        } else {
            return reservationTimeConfigurationService.findByDefaultConfigTrue();
        }
    }
}
*/

File: next-main/src/main/kotlin/com/next/reservations/core/service/ReservationTimeManagingService.kt
package com.next.reservations.core.service

import com.next.reservations.core.domain.ReservationStatus
import com.next.reservations.core.domain.ReservationTime
import com.next.reservations.core.domain.ReservationTimeConfiguration
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.util.stream.Collectors
import javax.transaction.Transactional


@Service
class ReservationTimeManagingService(private val reservationTimeService: ReservationTimeService,
                                     private val reservationTimeConfigurationService: ReservationTimeConfigurationService,
                                     private val reservationService: ReservationService,
                                     private val reservationManagingService: ReservationManagingService) {

    fun findUnreservedTimesForDate(date: LocalDate, timeConfiguration: ReservationTimeConfiguration): List<ReservationTime> {

        val reservedTimeIds = reservationService.findAllByStatusInAndReservationDate(listOf(ReservationStatus.ACCEPTED, ReservationStatus.PENDING), date)
                .stream().map { it.reservationTime!!.id }.collect(Collectors.toList())

        val availableTimes = reservationTimeService.findAllByReservationTimeConfigId(timeConfiguration.id)

        return availableTimes.stream().filter { !reservedTimeIds.contains(it.id) }.collect(Collectors.toList())
    }

    fun findConfigurationForDate(date: LocalDate): ReservationTimeConfiguration {

        val futureDefault: ReservationTimeConfiguration? = reservationTimeConfigurationService.findFutureDefault()

        return if (futureDefault != null && !date.isBefore(futureDefault.defaultStartDate))
            futureDefault
        else
            reservationTimeConfigurationService.findByDefaultConfigTrue()

    }

    @Transactional
    fun deleteReservationTimeConfig(id: Long) {
        reservationTimeConfigurationService.validateCanDelete(id)

        val reservationTimes = reservationTimeService.findAllByReservationTimeConfigId(id)

        reservationTimes.forEach { deleteReservationTime(it.id) }

        reservationTimeConfigurationService.deleteById(id)
    }

    @Transactional
    fun deleteReservationTime(reservationTimeId: Long) {
        val reservationsForRemoval = reservationManagingService.findAllActiveReservationsByReservationTime(reservationTimeId)
                .map { it.id }

        reservationService.removeReservationsByIds(reservationsForRemoval)

        reservationTimeService.delete(reservationTimeId)
    }

}
